# Analyzing Used Car Listings on eBay Kleinanzeigen
We will be working on a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website. 

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The dataset has 50,000 data points, and was modified by Dataquest to emulate scraped data with unclean data entries.

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
# We will import the libraries we will be using in our cleaning and analysis processes.
import pandas as pd
import numpy as np

#We will next open the 'auto.csv' file, and determine the type of encoding it uses.
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,odometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50


We can see there are 20 columns in our dataset, most of which are `object` Dtypes - stored as strings. There are some columns, such as `vehicleType`, which contain null values; `notRepairedDamage` has the most, with 9829 null values.

To first start our cleaning process, we'll clean the names of the columns.
## Cleaning Columns

In [2]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We will make the following changes:
- Change the columns from camelcase to snakecase
- Rename some columns, to make them more accurate descriptions

In [3]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

,date_crawled,name,seller,offer_type,price,ab_test,vehicle_type,registration_year,gearbox,power_ps,model,odometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,num_photos,postal_code,last_seen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50


The changes we have made make the data more readable and follow standard python convention for variable naming.
## Exploring and Cleaning
#### `seller`, `offer_type`, and `num_photos`
By exploring, we can find areas of the data we can easily clean.

In [4]:
autos.describe(include='all')

,date_crawled,name,seller,offer_type,price,ab_test,vehicle_type,registration_year,gearbox,power_ps,model,odometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,num_photos,postal_code,last_seen
count,50000,50000,50000,50000,50000,50000,44905,50000.000000,47320,50000.000000,47242,50000,50000.000000,45518,50000,40171,50000,50000.0,50000.000000,50000
unique,48213,38754,2,2,2357,2,8,NaN,2,NaN,245,13,NaN,7,40,2,76,NaN,NaN,39481
top,2016-04-04 16:40:33,Ford_Fiesta,privat,Angebot,$0,test,limousine,NaN,manuell,NaN,golf,"150,000km",NaN,benzin,volkswagen,nein,2016-04-03 00:00:00,NaN,NaN,2016-04-07 06:17:27
freq,3,78,49999,49999,1421,25756,12859,NaN,36993,NaN,4024,32424,NaN,30107,10687,35232,1946,NaN,NaN,8
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2005.073280,NaN,116.355920,NaN,NaN,5.723360,NaN,NaN,NaN,NaN,0.0,50813.627300,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,105.712813,NaN,209.216627,NaN,NaN,3.711984,NaN,NaN,NaN,NaN,0.0,25779.747957,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1000.000000,NaN,0.000000,NaN,NaN,0.000000,NaN,NaN,NaN,NaN,0.0,1067.000000,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1999.000000,NaN,70.000000,NaN,NaN,3.000000,NaN,NaN,NaN,NaN,0.0,30451.000000,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003.000000,NaN,105.000000,NaN,NaN,6.000000,NaN,NaN,NaN,NaN,0.0,49577.000000,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2008.000000,NaN,150.000000,NaN,NaN,9.000000,NaN,NaN,NaN,NaN,0.0,71540.000000,NaN


We can immediately see there are some odd columns:
- Both `seller` and `offer_type` show only 2 unique data entries, with all but 1 being the same.
- The `num_photos` column also appears odd, as it is a numerical column with apparantly no difference in data between entries. We will investigate this further to verify.

In [5]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

We can see the value `0` is given in all the data fields. We can see that the three columns we mentioned provide no useful data for analysis, so we will start by removing them from the dataset.

In [6]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

#### `price` and `odometer`
We can see that these columns are numerical values, which have extra characters - currency and distance measurements. We can remove them and rename the column to allow us to more easily mathematically manipulate the data.

In [7]:
autos["price"] = (autos["price"]
                             .str.replace(",", "")
                             .str.replace("$", "")
                             .astype(int)
                 )

autos["odometer"] = (autos["odometer"]
                             .str.replace(",", "")
                             .str.replace("km", "")
                             .astype(int)
                    )

autos.rename({"odometer": "odometer_km"}, axis = 1, inplace = True)
autos[["price", "odometer_km"]].head()

,price,odometer_km
0,5000,150000
1,8500,150000
2,8990,70000
3,4350,70000
4,1350,150000


## Exploring `price` and `odometer_km`
Having now cleaned the data, we can further explore the data and determine if we need to further remove/ clean the dataset.

In [8]:
print("Shape:")
print(autos["price"].unique().shape)
print()
print("Describe:")
print(autos["price"].describe())
print()
print("Value Count:")
autos["price"].value_counts().head(15)

Shape:
(2357,)

Describe:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Value Count:


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64

##### Prelimary Analysis
- We can see there are 2357 unique values.
- We can see we have many entries at a value of 0. We will most likely remove these data entries, since they only make a small proportion of the entire dataset.
- We have a maximum price of about $100,000,000. This is much too high to be a regular data entry, we will class it as an outlier. We will further explore the data and determine an appropriate cut-off for outliers, then remove them from the dataset as well.

In [9]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

From our analysis, we can see there are 14 cars in the dataset sold at extreme values of around `$1 million` or more. We will remove these from our dataset. We will also remove data entires with a price of `$0`. We will retain `$1` and above vehicles, as eBay is an auction site, so cars may genuinely be listed at `$1`.

In [10]:
autos = autos[autos["price"].between(1, 351000)]

## Exploring the date columns
We will be exploring the following columns:
- `date_crawled`
- `registration_month`
- `registration_year`
- `ad_created`
- `last_seen`

They contain date information, including when they were crawled. The non-registration dates are stored as strings, and we can explore them.

In [11]:
autos[['date_crawled','ad_created','last_seen']].head()

,date_crawled,ad_created,last_seen
0,2016-03-26 17:47:46,2016-03-26 00:00:00,2016-04-06 06:45:54
1,2016-04-04 13:38:56,2016-04-04 00:00:00,2016-04-06 14:45:08
2,2016-03-26 18:57:24,2016-03-26 00:00:00,2016-04-06 20:15:37
3,2016-03-12 16:58:10,2016-03-12 00:00:00,2016-03-15 03:16:28
4,2016-04-01 14:38:50,2016-04-01 00:00:00,2016-04-01 14:38:50


#### `date_crawled`

In [12]:
(autos["date_crawled"]
                .str[:10]
                .value_counts(normalize = True, dropna = False)
                .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Looking at the data, sorted by date, we can see the data was crawled for a roughly 1 month period between March 2016 to April 2016. We can see the distribuition is roughly uniform across each day.
#### `ad_created`

In [13]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of dates, there are 76 different dates. We can see most are created between 1 - 2 months between, but there are a few old ads, created up to 9 months ago.
#### `last_seen`

In [14]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler also records when a vehicle was last seen on a site. This implies the listing was removed, either by the owner, or it was sold. We can see the last few days particularly have large number of entries, this is more likely to be due to the fact the crawler was stopped, hence it was the last day recorded.
### Incorrect Entries for `registration_year`

In [15]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

This data should represent when the car was manufactured, however we can see there are outliers on either end. There are some cars registered as having been manufactured in 1000, and some in 9999. We have to deal with them. 

We immediately know any vehicle registering as beyond April 2016 is innacurate, as they would be from beyond when the listing was created. Determining the lower bounds is more challenging without further data. We can assume it would be in the early 1900s, when cars were being prodcued.

In [16]:
autos = autos[autos["registration_year"].between(1900, 2016)]
autos = autos[ ~ ((autos["registration_year"] == 2016) & (autos["registration_month"] > 4))]
autos["registration_year"].value_counts(normalize = True).head(10)

2000    0.068515
2005    0.063739
1999    0.062892
2004    0.058681
2003    0.058594
2006    0.057964
2001    0.057226
2002    0.053970
1998    0.051299
2007    0.049432
Name: registration_year, dtype: float64

We can see the most recent cars were only recently registered in the last 22 years.
## Exploring Price by Brand

In [17]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.210755
bmw               0.110609
opel              0.107092
mercedes_benz     0.096802
audi              0.086903
ford              0.069709
renault           0.047088
peugeot           0.029764
fiat              0.025595
seat              0.018127
skoda             0.016499
nissan            0.015197
mazda             0.015153
smart             0.014068
citroen           0.013937
toyota            0.012722
hyundai           0.010051
sonstige_autos    0.009899
volvo             0.009226
mini              0.008792
mitsubishi        0.008206
honda             0.007837
kia               0.007164
alfa_romeo        0.006621
porsche           0.006165
suzuki            0.005948
chevrolet         0.005710
chrysler          0.003495
dacia             0.002627
daihatsu          0.002518
jeep              0.002279
subaru            0.002149
land_rover        0.002106
saab              0.001672
jaguar            0.001585
daewoo            0.001520
trabant           0.001411
r

We can see there are many brands out there, however many don't have many listings. We will limit our exploration to brands with at least a 5% market presence.

In [18]:
brand_percentage = autos["brand"].value_counts(normalize = True)
brand_with_presence = brand_percentage[brand_percentage > 0.05].index
mean_prices = {}

for brand in brand_with_presence:
    branded_cars = autos[autos["brand"] == brand]
    mean_price = branded_cars["price"].mean()
    mean_prices[brand] = int(mean_price)

mean_prices

{'volkswagen': 5444,
 'bmw': 8367,
 'opel': 2994,
 'mercedes_benz': 8681,
 'audi': 9387,
 'ford': 3790}

We can see there is a distinct price gap, in the most frequently listed brands:
- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

In [19]:
bmp_series = pd.Series(mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

,mean_price
volkswagen,5444
bmw,8367
opel,2994
mercedes_benz,8681
audi,9387
ford,3790


## Exploring Mileage

In [20]:
mean_mileage = {}

for brand in brand_with_presence:
    branded_cars = autos[autos["brand"] == brand]
    mean_miles = branded_cars["odometer_km"].mean()
    mean_mileage[brand] = int(mean_miles)

bmm_series = pd.Series(mean_mileage)
pd.DataFrame(bmm_series, columns=["mean_mileage"])

,mean_mileage
volkswagen,128550
bmw,132490
opel,129195
mercedes_benz,130638
audi,129056
ford,124051


In [27]:
brand_mean_mileage = pd.Series(mean_mileage).sort_values(ascending=False)
brand_info = pd.DataFrame(brand_mean_mileage,columns=['mean_mileage'])

brand_mean_prices = pd.Series(mean_prices).sort_values(ascending=False)
brand_info["mean_price"] = brand_mean_prices 

brand_info

,mean_mileage,mean_price
bmw,132490,8367
mercedes_benz,130638,8681
opel,129195,2994
audi,129056,9387
volkswagen,128550,5444
ford,124051,3790


The range of car mileages does not vary as much as the prices do by brand. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.